graylog-storage-elasticsearch7/src/main/java/org/graylog/storage/elasticsearch7/views/searchtypes/pivot/ESPivot.java

```diff
@@ -91,82 +93,79 @@ public void doGenerateQueryPart(SearchJob job, Query query, Pivot pivot, ESGener
-                            .forEach(aggregationBuilder::subAggregation);
-                }
-                if (previousAggregation != null) {
-                    previousAggregation.subAggregation(aggregationBuilder);
-                } else {
-                    searchSourceBuilder.aggregation(aggregationBuilder);
-                }
-                previousAggregation = aggregationBuilder;
-            }
+
+        final Iterator<BucketSpec> columnBuckets = pivot.columnGroups().iterator();
+        while (columnBuckets.hasNext()) {
+            final BucketSpec bucketSpec = columnBuckets.next();
+            final boolean isLastColumnBucket = !columnBuckets.hasNext();
+            final Optional<AggregationBuilder> generateSingleBucketAggregation = doGenerateSingleBucketAggregation(bucketSpec, isLastColumnBucket, "column", query, pivot, queryContext);
+            generateSingleBucketAggregation.ifPresent(bucketAggregationChain::addFirst);
         }
 
-        final MinAggregationBuilder startTimestamp = AggregationBuilders.min("timestamp-min").field("timestamp");
-        final MaxAggregationBuilder endTimestamp = AggregationBuilders.max("timestamp-max").field("timestamp");
+        return bucketAggregationChain.stream()
+                .reduce((aggrLower, aggrHigher) -> {
+                    aggrHigher.subAggregation(aggrLower);
+                    return aggrHigher;
+                });
 
-        searchSourceBuilder.aggregation(startTimestamp);
-        searchSourceBuilder.aggregation(endTimestamp);
+    }
 
-        if (topLevelAggregation == null) {
-            LOG.debug("No aggregations generated for {}", pivot);
+    private Optional<AggregationBuilder> doGenerateSingleBucketAggregation(BucketSpec bucketSpec,
+                                                                           final boolean isLast,
+                                                                           final String reason,
+                                                                           final Query query,
+                                                                           Pivot pivot,
+                                                                           ESGeneratedQueryContext queryContext
+    ) {
+        final String name = queryContext.nextName();
+        LOG.debug("Creating " + reason + " group aggregation '{}' as {}", bucketSpec.type(), name);
+        final ESPivotBucketSpecHandler<? extends PivotSpec, ? extends Aggregation> handler = bucketHandlers.get(bucketSpec.type());
+        if (handler == null) {
+            throw new IllegalArgumentException("Unknown " + reason + "_group type " + bucketSpec.type());
+        }
+        final Optional<AggregationBuilder> generatedAggregation = handler.createAggregation(name, pivot, bucketSpec, this, queryContext, query);
+        if (generatedAggregation.isPresent()) {
```

Comment:
why not like ?
`        generatedAggregation.ifPresent(aggregationBuilder -> {
            if (isLast || pivot.rollup()) {
                seriesStream(pivot, queryContext, isLast ? "leaf " + reason : reason + " rollup")
                        .forEach(aggregationBuilder::subAggregation);
            }
        });
`